### ISSUE TITLE
"Finished" run changing its state to "started" when new machine is joined after finish

### ISSUE THREAD

**Issue by denis-domanskii** [sorry-cypress/sorry-cypress#215]
## How to reproduce

1) Start parallel run on the first machine
2) Wait until run is finished
3) Start parallel run with same build-id on the second machine. It is is real-life case because second machine was busy by other task for a long time and started Cypress task only when the run was already finished. 

## Expected
Second machine doesn't trigger "RUN_START" event for already finished run.

## Observed
Second machine triggers "RUN_START" event for already finished run. This events triggers GitHub Status event handler, and incorrect status "PENDING" is sent to GitHub. Also there is no second "RUN_FINISH" event, therefore GutHub status is not finalised at all.

## Environment

- sorry-cypress version: `v1.0.0-beta.10`
- cypress version: `6.0.0`
- platform: `Docker`
- service: `api | director`

## Logs
See runId `08032e5fb82f125e614f20a2198f461e`:
```
>> Received instance result { instanceId: 'ccbbb38c-b9fc-4917-afcf-d763b0313435' }
<< INSTANCE_FINISH hook called ccbbb38c-b9fc-4917-afcf-d763b0313435
<< RUN_FINISH hook called 08032e5fb82f125e614f20a2198f461e
<< Sending assets upload URLs {
  instanceId: 'ccbbb38c-b9fc-4917-afcf-d763b0313435',
  screenshotUploadUrls: [],
  videoUploadInstructions: null
}
>> [not implemented] Received stdout for instance { instanceId: 'ccbbb38c-b9fc-4917-afcf-d763b0313435' }
>> Machine is requesting a new task {
  runId: '08032e5fb82f125e614f20a2198f461e',
  machineId: '48636144-339b-42db-83e4-9ce33eff1c40',
  groupId: '24b90b84f4489c2b850c67e9acd73a26dc9ab679'
}
<< All tasks claimed {
  runId: '08032e5fb82f125e614f20a2198f461e',
  machineId: '48636144-339b-42db-83e4-9ce33eff1c40'
}
>> Machine is joining / creating  a run { ciBuildId: '24b90b84f4489c2b850c67e9acd73a26dc9ab679', group: null }
<< RUN_START hook called {
  groupId: '24b90b84f4489c2b850c67e9acd73a26dc9ab679',
  machineId: '45c07085-5127-4c78-91e5-c1563ab52a67',
  runId: '08032e5fb82f125e614f20a2198f461e',
  runUrl: 'http://qa.server.net/run/08032e5fb82f125e614f20a2198f461e',
  warnings: []
}
<< Responding to machine {
  groupId: '24b90b84f4489c2b850c67e9acd73a26dc9ab679',
  machineId: '45c07085-5127-4c78-91e5-c1563ab52a67',
  runId: '08032e5fb82f125e614f20a2198f461e',
  runUrl: 'http://qa.server.net:8080/run/08032e5fb82f125e614f20a2198f461e',
  warnings: []
}
>> Machine is requesting a new task {
  runId: '08032e5fb82f125e614f20a2198f461e',
  machineId: '45c07085-5127-4c78-91e5-c1563ab52a67',
  groupId: '24b90b84f4489c2b850c67e9acd73a26dc9ab679'
}
<< All tasks claimed {
  runId: '08032e5fb82f125e614f20a2198f461e',
  machineId: '45c07085-5127-4c78-91e5-c1563ab52a67'
}
```

**Comment by agoldis** [comment]
Somewhat related: 

- https://github.com/sorry-cypress/sorry-cypress/issues/134
- https://github.com/sorry-cypress/sorry-cypress/issues/142

Ideally, we'd love to have a 'grace period' for runs, when expired, new machines won't be allowed and non-finished instances would time out. 

From the original cypress:
https://github.com/cypress-io/cypress/issues/1891

**Comment by denis-domanskii** [comment]
@agoldis probably we just shouldn't create a new run if a run with the same id is already finished? If new run will not be created - RUN_START hook will not be raised. A machine which trying to connect with finished run id - should receive "All tasks claimed" response.

**Comment by agoldis** [comment]
> probably we just shouldn't create a new run if a run with the same id is already finished

Yes! The challenge is to detect when the run is finished.

**Comment by denis-domanskii** [comment]
> Yes! The challenge is to detect when the run is finished.

Probably I'm not familiar enough with a cypress/sorry-cypress protocol, but as I understand - sorry-cypress knows total specs count, sorry-cypress knows each spec status (and shows it in a report), therefore run is finished when all specs are finished. In this moment we can set flag "finished" for specific run and don't try to start it again when a new machine is joined.

**Comment by agoldis** [comment]
TL;DR We don't really know all the specs upfront, we dynamically add specs when a new machine joins. The only solution I can see it some kind of deadline for adding new specs to a run.

--- 

E.g. given two groups of specs and 4 machines:

```
groupA
 - spec A1
 - spec A2

groupB
- spec B1
- spec B2
- spec B3
```

1. machine 1 runs groupA and creates a run populates specs for group A, gets spec A1
2. machine 2 runs groupA, gets spec A2

3. 3 minutes later, machine 3 joins the run with group B specs, we add group B to the run, the machines starts running spec B1
4. machine 4 joins and starts running B2...


It shouldn't be a huge deal, it's just I don't have enough free time to tackle that.

**Comment by denis-domanskii** [comment]
Ok, got it. But probably, to be consistent, we should trigger RUN_FINISH event when last known spec is finished?

Because now the flow is:
- ...
- _[run in progress]_
- last spec is executed
- INSTANCE_FINISH hook called
- RUN_FINISH hook called
- _[run is finished]_
- new machine is joined with the same specs set and same ci-build-id
- RUN_START hook called
- _[run in progress]_
- "All tasks claimed" message is sent to machine
- _[run in progress forever]_

I propose to adjust it by adding:
- ...
- _[run in progress]_
- last spec is executed
- INSTANCE_FINISH hook called
- RUN_FINISH hook called
- _[run is finished]_
- new machine is joined with the same specs set and same ci-build-id
- RUN_START hook called
- _[run in progress]_
- "All tasks claimed" message is sent to machine
- **INSTANCE_FINISH hook called**
- **RUN_FINISH hook called**
- **[run is finished]**

It's important because without second "RUN_FINISH" hook - status in GitHub hangs as "Pending". And if this check is required to merge changes, this merge will be blocked.

**Comment by agoldis** [comment]
You're right that's quite important. I will try to resolve this and #134 altogether.

**Comment by agoldis** [comment]
Resolved in https://github.com/sorry-cypress/sorry-cypress/pull/292
